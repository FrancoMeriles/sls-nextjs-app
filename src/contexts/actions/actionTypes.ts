export const ADD_FAVORITE = 'ADD_FAVORITE'
export const REMOVE_FAVORITE = 'REMOVE_FAVORITE'
export const BATCH_FAVORITES = 'BATCH_FAVORITES'

export const SET_ZIPCODE = 'SET_ZIPCODE'
