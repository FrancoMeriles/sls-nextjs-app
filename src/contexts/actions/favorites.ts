import * as actionTypes from './actionTypes'

export const addFavorite = (product) => {
  return { type: actionTypes.ADD_FAVORITE, product: product }
}

export const removeFavorite = (id) => {
  return { type: actionTypes.REMOVE_FAVORITE, id: id }
}

export const batchFavorites = (products) => {
  return { type: actionTypes.BATCH_FAVORITES, products: products }
}
