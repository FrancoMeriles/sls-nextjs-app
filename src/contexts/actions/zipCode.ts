import * as actionTypes from './actionTypes'

export interface ZipCodeActions {
  type: string
  zipCode: number
}

export const setZipCode = (zipCode: number): ZipCodeActions => {
  return { type: actionTypes.SET_ZIPCODE, zipCode: zipCode }
}
