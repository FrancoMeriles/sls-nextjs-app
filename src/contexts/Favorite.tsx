import { useReducer, useContext, createContext, useEffect, useState } from 'react'
import * as actionTypes from './actions/actionTypes'
import * as favoriteAction from './actions/favorites'

const addFavorite = (state, action) => [...state, action.product]
const batchFavorites = (state, action) => [...state, ...action.products]
const removeFavorite = (state, action) => state.filter((product) => product.id !== action.id)

const FavoriteStateContext = createContext(null)
const FavoriteDispatchContext = createContext(null)

const reducer = (state, action) => {
  switch (action.type) {
    case actionTypes.ADD_FAVORITE: {
      const newState = addFavorite(state, action)
      localStorage.setItem('favoriteProducts', JSON.stringify(newState))
      return newState
    }
    case actionTypes.REMOVE_FAVORITE: {
      const newState = removeFavorite(state, action)
      localStorage.setItem('favoriteProducts', JSON.stringify(newState))
      return newState
    }
    case actionTypes.BATCH_FAVORITES:
      return batchFavorites(state, action)
    default:
      return state
  }
}

export const FavoriteProvider = ({ children }) => {
  const [initialState, setInitialState] = useState([])
  const [state, dispatch] = useReducer(reducer, initialState)

  useEffect(() => {
    const favoriteProductsLocalStorage = localStorage.getItem('favoriteProducts')
    if (favoriteProductsLocalStorage) {
      dispatch(favoriteAction.batchFavorites(JSON.parse(favoriteProductsLocalStorage)))
      setInitialState(JSON.parse(favoriteProductsLocalStorage))
    }
  }, [])

  return (
    <FavoriteDispatchContext.Provider value={dispatch}>
      <FavoriteStateContext.Provider value={state}>{children}</FavoriteStateContext.Provider>
    </FavoriteDispatchContext.Provider>
  )
}

export const useFavorite = () => useContext(FavoriteStateContext)
export const useDispatchFavorite = () => useContext(FavoriteDispatchContext)
