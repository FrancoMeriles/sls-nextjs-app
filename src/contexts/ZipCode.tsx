import { useReducer, useContext, createContext, useEffect, useState } from 'react'
import ZipCodeService from '@base/service/zipcode.service'
import * as actionTypes from './actions/actionTypes'
import * as zipCodeAction from './actions/zipCode'

const ZipCodeStateContext = createContext(null)
const ZipCodeDispatchContext = createContext(null)

const reducer = (state, action) => {
  switch (action.type) {
    case actionTypes.SET_ZIPCODE: {
      const newZipCode = action.zipCode
      localStorage.setItem('zipCode', action.zipCode)
      return newZipCode
    }
    default:
      return state
  }
}

export const ZipCodeProvider = ({ children }) => {
  const [initialState, setInitialState] = useState(1013)
  const [state, dispatch] = useReducer(reducer, initialState)
  useEffect(() => {
    const zipCodeLocalStorage = localStorage.getItem('zipCode')
    const getZipCode = async () => {
      const zipCodeApi = new ZipCodeService()
      const zipCode = await zipCodeApi.getZipCode()
      dispatch(zipCodeAction.setZipCode(Number(zipCode.postal_code) || initialState))
      setInitialState(Number(zipCode.postal_code))
    }
    getZipCode()
    if (zipCodeLocalStorage) {
      dispatch(zipCodeAction.setZipCode(zipCodeLocalStorage))
    }
  }, [])

  return (
    <ZipCodeDispatchContext.Provider value={dispatch}>
      <ZipCodeStateContext.Provider value={state}>{children}</ZipCodeStateContext.Provider>
    </ZipCodeDispatchContext.Provider>
  )
}

export const useZipCode = () => useContext(ZipCodeStateContext)
export const useDispatchZipCode = () => useContext(ZipCodeDispatchContext)
