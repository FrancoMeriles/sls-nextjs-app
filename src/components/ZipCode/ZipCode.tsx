import React from 'react'
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
  Input,
  PopoverBody,
  PopoverArrow,
  PopoverCloseButton,
  Text,
  InputGroup,
  InputRightElement,
  Button,
  useDisclosure,
} from '@chakra-ui/react'
import { useZipCode, useDispatchZipCode } from '@base/contexts/ZipCode'
import { setZipCode } from '@base/contexts/actions/zipCode'

import { Position } from '@base/icons'

const ZipCode = () => {
  const { isOpen, onOpen, onClose } = useDisclosure()
  const zipCode = useZipCode()
  const dispatch = useDispatchZipCode()
  const [value, setValue] = React.useState('')

  const handleChange = (e: React.FormEvent<EventTarget>): void => {
    const target = e.target as HTMLInputElement
    setValue(target.value)
  }
  const handleFormSubmit = (e: React.FormEvent<EventTarget>): void => {
    e.preventDefault()
    if (value) {
      dispatch(setZipCode(value))
    }
    onClose()
  }

  return (
    <Popover isOpen={isOpen} onOpen={onOpen} onClose={onClose}>
      <PopoverTrigger>
        <Text
          fontSize="xs"
          fontWeight="500"
          color="#585858"
          ml={2}
          cursor="pointer"
          _hover={{ textDecoration: 'underline' }}
        >
          <Position boxSize={3} mb="2px" mr={1} />
          {zipCode && `CP: ${zipCode}`}
        </Text>
      </PopoverTrigger>
      <PopoverContent>
        <PopoverArrow />
        <PopoverCloseButton />
        <PopoverBody p="30px 15px 15px 15px">
          <form onSubmit={handleFormSubmit}>
            <InputGroup borderColor="transparent">
              <Input
                value={value}
                onChange={handleChange}
                variant="outline"
                placeholder="Zipcode"
                size="lg"
                focusBorderColor="gray.100"
                backgroundColor="white"
              />
              <InputRightElement
                width="auto"
                pr={4}
                height="100%"
                children={
                  <Button type="submit" height="100%" variant="link">
                    Cambiar
                  </Button>
                }
              />
            </InputGroup>
          </form>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  )
}

export default ZipCode
