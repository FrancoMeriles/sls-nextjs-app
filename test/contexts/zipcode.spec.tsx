import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'
// const { act } = renderer

import { ZipCodeProvider, useZipCode, useDispatchZipCode } from '@base/contexts/ZipCode'
import { setZipCode } from '@base/contexts/actions/zipCode'

const mockZipCode = {
  postal_code: 2000,
}

jest.mock(
  '@base/service/zipcode.service',
  () =>
    class MockZipCode {
      async getZipCode() {
        return Promise.resolve(mockZipCode)
      }
    }
)

const DumbComponent = () => {
  const zipcode = useZipCode()
  const dispatch = useDispatchZipCode()
  return (
    <div>
      {zipcode}
      <button onClick={() => dispatch(setZipCode(5500))}>Set ZipCode</button>
      <button onClick={() => dispatch({ type: 'UNDEFINDED_ACTION' })}>
        Call for undefined action
      </button>
    </div>
  )
}

describe('Favorite Context', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    jest.clearAllMocks()
  })
  it('should render and match snapshot with required children', async () => {
    const tree = renderer
      .create(
        <ZipCodeProvider>
          <div>__CHILDREN__</div>
        </ZipCodeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should set a zipcode', async () => {
    await act(async () =>
      render(
        <ZipCodeProvider>
          <DumbComponent />
        </ZipCodeProvider>
      )
    )
    expect(screen.getByText('2000')).toBeInTheDocument()
    userEvent.click(screen.getByText('Set ZipCode'))
    expect(screen.getByText('5500')).toBeInTheDocument()
  })

  it('should detect a zipcode in localstoraeset a zipcode', async () => {
    const mockGetItem = jest.fn(() => '2500')
    Storage.prototype.getItem = mockGetItem

    await act(async () =>
      render(
        <ZipCodeProvider>
          <DumbComponent />
        </ZipCodeProvider>
      )
    )
    expect(screen.getByText('2000')).toBeInTheDocument()
    userEvent.click(screen.getByText('Call for undefined action'))
    expect(screen.getByText('2000')).toBeInTheDocument()
  })
})
