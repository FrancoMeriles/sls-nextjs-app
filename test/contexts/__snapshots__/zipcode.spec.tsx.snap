// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite Context should render and match snapshot with required children 1`] = `
<div>
  __CHILDREN__
</div>
`;
